### PR TITLE
Backend: Avoid temporary variable in `set_val`

### PR DESCRIPTION
**Pull request by nomeata**
as suggested by @ggreif. Fixes: #3402

### CODE REVIEW

**Comment by nomeata** [comment]
Something is still wrong:
```
--- actor-class-arg.drun (expected)
+++ actor-class-arg.drun (actual)
@@ -20,8 +20,7 @@
 debug.print: Hello
 debug.print: World
 ingress Completed: Reply: 0x4449444c0000
-debug.print: Hello
-ingress Completed: Reply: 0x4449444c0000
+ingress Err: IC0502: Canister rwlgt-iiaaa-aaaaa-aaaaa-cai trapped: stack overflow
 Ok: Reply: 0x4449444c0000
 debug.print: Hello
 debug.print: World
```

**Comment by nomeata** [comment]
What is the difference between 
```
actor class (t : Text, t1 : Text) {
```
and
```
shared actor class (t : Text, t1 : Text) {
```
?

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -6770,11 +6770,24 @@ module Var = struct
     | Some (PublicMethod _) -> fatal "set_val: %s is PublicMethod" var
     | None   -> fatal "set_val: %s missing" var
 
-  (* Stores the payload (which is found on the stack, in Vanilla stackrep) *)
+  (* Stores the payload (using preparation code and consumptiong code) *)
```
Comment (ggreif) [inline src/codegen/compile.ml:6773]: consuming? downstream?

Comment (nomeata) [reply src/codegen/compile.ml:6773]:
```suggestion
  (* Stores the payload. Returns stack preparation code, and code that consumes the values from the stack *)
```

**Comment by nomeata** [comment]
So much for premature optimization:

>  In terms of size, 4 tests improved and the mean change is -0.1%.

**Comment by crusso** [comment]
> What is the difference between
> 
> ```
> actor class (t : Text, t1 : Text) {
> ```
> 
> and
> 
> ```
> shared actor class (t : Text, t1 : Text) {
> ```
> 
> ?
Six characters? 

The latter gives you the option of accessing the caller...
```
 shared {caller} actor class (t : Text, t1 : Text) {
```

**Comment by crusso** [comment]
> > What is the difference between
> > ```
> > actor class (t : Text, t1 : Text) {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```
> > shared actor class (t : Text, t1 : Text) {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ?
> 
> Six characters?
> 
> The latter gives you the option of accessing the caller...
> 
> ```
>  shared {caller} actor class (t : Text, t1 : Text) {
> ```

Hmm, maybe we should have attached it to `actor` and `func` (as an optional prefix) instead, so you could get the owner of a plain old actor without having to declare a class....

**Comment by nomeata** [comment]
I was just surprised it changed behaviour somehow – with `shared` it triggered an earelier bug on this PR, without it didn’t

**Review by ggreif — COMMENTED** [review]
Is there any change in the evaluation order?

**Comment by nomeata** [comment]
No, it's only moving the address pushes earlier, no side effects there

**Comment by crusso** [comment]
> No, it's only moving the address pushes earlier, no side effects there 

It might have to move the context pat into the scope of the arguments,  requiring some renaming and extra code. Maybe the compiler doesn't spot that the pat is still trivial in this case.

**Comment by nomeata** [comment]
I don't follow. I merely reorder some wasm instructions. Nothing big source-levely happening.

**Review by ggreif — APPROVED** [review]
LGTM! Thanks for tackling this!

**Pull request merged**